### DISCUSSION TITLE
Access response body of errors

### DISCUSSION THREAD

**Discussion by dios-david** [urql-graphql/urql#2041] [Q&A]
I wanted to access the body of errors responses in a custom exchange, but I couldn't make it happen.
The issue is that the body of the Response was already read by urql, so I can't do `operationResult.error.response.json()` as Response body can only read once :/

I found an attempt from earlier which tried to utilise `response.clone()`, but that was closed due to browser compatibility: https://github.com/FormidableLabs/urql/pull/952

Is there any alternative solution for this?

**Comment by dios-david** [comment]
The only solution I can think of is passing in a custom `fetch` implementation to `createClient()`, but I would prefer to do this in an exchange. 🤔

**Reply by JoviDeCroock** [reply]
An option would be to write your own `fetchExchange`, what is the use case in specific? Because normally both `errors` and `data` should be translated even with errors

**Reply by dios-david** [reply]
I have a bot detection app running on the graphql endpoint, which returns a 403 response with some data in the response if it detects the user as a bot. The data in the response then needs to be passed in to a client sdk in the browser (to do some tracking and triggering a captcha page etc), that's why I need to access the response body.

**Response body**
![Screenshot 2021-10-22 at 17 11 24](https://user-images.githubusercontent.com/3816644/138488179-4bbdc1ba-5d01-4937-955f-97d1e985809e.png)
(There is a `content-type: application/json; charset=utf-8` response header as well)

**Operation logged in the exchange**
![Screenshot 2021-10-22 at 17 10 49](https://user-images.githubusercontent.com/3816644/138488181-c5168989-91eb-4d66-b95d-d997906c4af7.png)

The exchange I used here is this:
```tsx
import { Exchange } from 'urql';
import { pipe, tap } from 'wonka';

export const myDemoExchange: Exchange = ({ forward }) => (ops$) => pipe(
  ops$,
  forward,
  tap((operationResult) => {
    console.log(operationResult);
  }),
);
```

**Reply by JoviDeCroock** [reply]
I wonder if this is better suited for `extensions` WDYT? https://github.com/FormidableLabs/urql/blob/922d4a110baa0cb4b0ea75734ef9d40fa403e194/packages/core/src/utils/result.ts#L23

**Reply by dios-david** [reply]
I totally agree with that, but it's a 3rd party so I can't change the response body to have `extensions` :(

**Reply by JoviDeCroock** [reply]
Darn 😅 yeah it's quite hard as it deviates from the specced response-format, I'd honestly say this is a case for a custom-fetch exchange
